primitives/src/credit.rs

```diff
@@ -0,0 +1,230 @@
+    }
+}
+
+impl From<u8> for CreditLevel {
+    fn from(num: u8) -> Self {
+        match num {
+            0 => Self::Zero,
+            1 => Self::One,
+            2 => Self::Two,
+            3 => Self::Three,
+            4 => Self::Four,
+            5 => Self::Five,
+            6 => Self::Six,
+            7 => Self::Seven,
+            8 => Self::Eight,
+            _ => Self::Zero,
+        }
+    }
+}
+
+impl Into<u8> for CreditLevel {
+    fn into(self) -> u8 {
+        match self {
+            Self::Zero => 0,
+            Self::One => 1,
+            Self::Two => 2,
+            Self::Three => 3,
+            Self::Four => 4,
+            Self::Five => 5,
+            Self::Six => 6,
+            Self::Seven => 7,
+            Self::Eight => 8,
+        }
+    }
+}
+
+impl CreditLevel {
+    pub fn get_credit_level(credit_score: u64) -> CreditLevel {
+        let credit_level = match credit_score {
+            0..=99 => CreditLevel::Zero,
+            100..=199 => CreditLevel::One,
+            200..=299 => CreditLevel::Two,
+            300..=399 => CreditLevel::Three,
+            400..=499 => CreditLevel::Four,
+            500..=599 => CreditLevel::Five,
+            600..=699 => CreditLevel::Six,
+            700..=799 => CreditLevel::Seven,
+            _ => CreditLevel::Eight,
+        };
+        credit_level
+    }
+
+    pub fn credit_level_gap(lhs: CreditLevel, rhs: CreditLevel) -> u64 {
+        let lhs: u8 = lhs.into();
+        let rhs: u8 = rhs.into();
+        100u64 * (lhs.saturating_sub(rhs) as u64)
+    }
+}
+
+pub trait CreditInterface<AccountId, Balance> {
+    fn get_credit_score(account_id: &AccountId) -> Option<u64>;
+    fn pass_threshold(account_id: &AccountId) -> bool;
+    fn slash_credit(account_id: &AccountId, score: Option<u64>) -> Weight;
+    fn get_credit_level(credit_score: u64) -> CreditLevel;
+    fn get_reward(
+        account_id: &AccountId,
+        from: EraIndex,
+        to: EraIndex,
+    ) -> (Option<(Balance, Balance)>, Weight);
+    fn get_top_referee_reward(account_id: &AccountId) -> (Balance, Weight);
+    fn update_credit_by_traffic(server: AccountId);
+    fn get_current_era() -> EraIndex;
+    fn update_credit_by_tip(who: AccountId, add_credit: u64);
+    fn update_credit_by_burn_nft(who: AccountId, add_credit: u64) -> DispatchResult;
+    fn init_delegator_history(account_id: &AccountId, era: u32) -> bool;
+    fn get_credit_balance(account_id: &AccountId) -> Vec<Balance>;
+    fn get_credit_gap(dst_lv: u8, cur_lv: u8) -> u64;
+    fn add_or_update_credit(account_id: AccountId, credit_score: u64);
+    fn is_first_campaign_end(account_id: &AccountId) -> Option<bool>;
+    fn do_unstaking_slash_credit(user: &AccountId) -> DispatchResult;
+    fn burn_record(burn_amount: Balance) -> bool;
+    fn get_credit_history(account_id: &AccountId) -> Vec<(EraIndex, CreditData)>;
+}
+
+impl<AccountId, Balance: From<u32>> CreditInterface<AccountId, Balance> for () {
```

Comment:
I think we can move it to `operation/src/tests.rs` because it's only used by test

Reply:
I think keep the default implement of thos interface is ok.
and you can't move the implement to other file